src/asmcup/genetics/Gene.java

```diff
@@ -10,16 +12,27 @@ public Gene(byte[] dna, int gen, float score) {
 		this.gen = gen;
 		this.score = score;
 	}
-	
+
+	@Override
 	public int compareTo(Gene other) {
 		float d = score - other.score;
 		
-		if (d == 0) {
+		if (Math.abs(d) < 0.00000001f) {
 			return 0;
 		} else if (d < 0) {
 			return 1;
 		}
 		
 		return -1;
 	}
+
+	@Override
+	public boolean equals(Object obj) {
+		return obj instanceof Gene && compareTo((Gene) obj) == 0;
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(score);

```

Comment:
Not quite. The default behavior (if I'm not mistaken) is that the object pointer is hashed (or something along those lines, point being that different [in the sense of "not the exact same instance"] objects usually have different hash codes and don't equal each other).

This means that the current implementation is [not consistent with equals](http://stackoverflow.com/questions/12123960/what-does-comparison-being-consistent-with-equals-mean-what-can-possibly-happe), but doesn't bog down e.g. a hash map.

The implication of this (to my understanding) is that (currently) asking a hash map for a `new Gene(score = z)` will return nothing, even though `new Gene(score = z).compareTo(g1)` would return true.

That's not a problem, because it's important to note that our genes must be sorted based on their score, but having the same score does not imply they are equal.